File: card-lang-typescript/src/types/Lexer.ts

import { Token } from "./ReusedStructs";

export class Lexer {
  current_tokenizer: string = "";

  input_chars: Array<string> = [];
  char_index: number = 0;
  tokens: Array<Token> = [];
  controller = new LexerController(this);

  constructor(
    private tokenizers: Array<Tokenizer>
  ){}
  reset(){
    this.current_tokenizer = "";
    this.input_chars = [];
    this.char_index = 0;
    this.tokens = [];
    return this;
  }
  tokenizeString(input_str: string): Array<Token>{
    this.reset();
    this.input_chars = Array.from(input_str);

    let tokenizers = this.tokenizers;
    var controller = this.controller;

    while(true){
      var char = controller.current_char;
      var usedTokenizer = false;
      if(typeof char == "undefined"){
        break;
      }
      for(let t of tokenizers){
        if(!t.matchesChar(char)){
          continue;
        }
        usedTokenizer = true;
        this.current_tokenizer = t.token_type;
        t.handleChar(char, controller);
        break;
      }
      if(usedTokenizer == false){
        throw new Error(`invalid token ${controller.current_char}`);
      }
    }
    return this.tokens;
  }
}

export class LexerController {
  constructor(private lexer: Lexer){}
  get current_char(): string | void {
    return this.lexer.input_chars[this.lexer.char_index];
  }
  get finished(){
    return this.lexer.char_index >= this.lexer.input_chars.length;    
  }
  advance(){
    this.lexer.char_index++
    return this.current_char;
  }
  addToken(values: {[key: string]: string}){
    this.lexer.tokens.push({
      token_type: this.lexer.current_tokenizer,
      values: values
    });
    return self;
  }
}

export interface Tokenizer {
  token_type: string;
  matchesChar(c: string): boolean;
  handleChar(c: string, lexer: LexerController): void;
}

// pub struct Tokenizer<Adv, Tok>{
//   token_type: String,
//   matchesChar: fn(c: char)->bool,
//   handleChar: fn(
//     initial_char: char,
//     advance: Adv,
//     addToken: Tok,
//   )->Result<(), String> where Adv: Fn()->char, Tok: Fn(HashMap<String, String>)->()
// }

// advance: impl Fn()->char,
// addToken: fn(values: HashMap<String, String>)->(),

File: card-lang-rust/src/types/Lexer.rs

use core::panic;
use std::{collections::HashMap, str::Chars};

use crate::types::ReusedStructs::Token;


pub struct Lexer<'a> {
  tokenizers: Vec<Box<dyn Tokenizer>>,
  current_tokenizer: String,
  input_chars: Chars<'a>,
  current_char: Option<char>,
  tokens: Vec<Token>
}

impl Lexer<'_> {
  pub fn new(tokenizers: Vec<Box<dyn Tokenizer>>)-> Lexer<'static> {
    return Lexer {
      tokenizers: tokenizers,
      current_tokenizer: "".to_string(),
      input_chars: "".chars(),
      current_char: None,
      tokens: vec![]
    };
  }
  fn reset(&self)->(){
    self.current_tokenizer = "".to_string();
    self.input_chars = "".chars();
    self.current_char = None;
    self.tokens = vec![];
  }
  pub fn advance(&self)->char{
    let c = self.input_chars.next();
    self.current_char = c;
    return c.unwrap();
  }
  pub fn addToken(&self, values: HashMap<String, String>)->(){
    self.tokens.push(Token {
      token_type: self.current_tokenizer.clone(),
      values: values
    });
  }
  pub fn tokenizeString(&self, input_str: String) -> Vec<Token>{
    let mut s = self.reset();
    let tokenizers = &self.tokenizers;
    self.input_chars = input_str.chars();
    self.advance();


    while self.current_char != None {
      let mut usedTokenizer = false;
      for t in tokenizers {
        if !t.matchesChar(self.current_char.clone().unwrap()) {
          continue;
        }
        usedTokenizer = true;
        self.current_tokenizer = t.token_type();
        if let Err(e) = t.handleChar(self.current_char.clone().unwrap(), self) {
          panic!("Error parsing the code");
        }
      }
      if usedTokenizer == false {
        panic!("invalid token {}", self.current_char.unwrap());
      }
    }
    return self.tokens;
  }

}

pub trait Tokenizer {
  fn token_type(&self)-> String;
  fn matchesChar(&self, c: char)->bool;
  fn handleChar(
    &self, c: char, lexer: &Lexer
  )->Result<(), String>;
}

// pub struct Tokenizer<Adv, Tok>{
//   token_type: String,
//   matchesChar: fn(c: char)->bool,
//   handleChar: fn(
//     initial_char: char,
//     advance: Adv,
//     addToken: Tok,
//   )->Result<(), String> where Adv: Fn()->char, Tok: Fn(HashMap<String, String>)->()
// }

// advance: impl Fn()->char,
// addToken: fn(values: HashMap<String, String>)->(),

